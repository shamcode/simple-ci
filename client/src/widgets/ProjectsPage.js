import { options } from 'sham-ui';
import ProjectsPageTemplate from './ProjectsPage.sht';

export default class ProjectsPage extends ProjectsPageTemplate {
    @options get dataLoaded() { return false;  }
    @options get projects() { return []; }
    @options get errors() { return []; }

    @options
    afterRender() {
        this._loadPageData();
    }

    _loadPageData() {
        const loading = new Promise( ( resolve ) => {
            setTimeout( () => resolve( {
                projects: [
                    { name: 'Fixture' },
                    { name: 'Second fixture' },
                    { name: 'Simple CI' },
                    { name: 'Simple CI 1' },
                    { name: 'Simple CI 2' },
                    { name: 'Snackbars are often used as tooltips/popups' }
                ]
            } ), 500 );
        } );
        loading.then(
            ::this._loadedPageDataSuccess,
            ::this._loadedPageDataFail
        ).finally(
            () => this.update( { dataLoaded: true } )
        )
    }

    _loadedPageDataSuccess( { projects } ) {
        this.update( {
            projects,
            errors: []
        } )
    }

    _loadedPageDataFail() {
        this.update( {
            projects: [],
            errors: [ "Fail loaded project!" ]
        } )
    }
}