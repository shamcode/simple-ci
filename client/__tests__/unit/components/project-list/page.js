import { createDI } from 'sham-ui';
import ProjectsPage from '../../../../src/components/project-list/page.sfc';
import * as directives from 'sham-ui-directives';
import hrefto from 'sham-ui-router/lib/href-to';
import renderer from 'sham-ui-test-helpers';

it( 'renders correctly', () => {
    const DI = createDI();
    const storeMock = jest.fn().mockReturnValue( Promise.resolve( { projects: [] } ) );
    DI.bind( 'store', {
        getProjects: storeMock
    } );
    DI.bind( 'router', {
        generate: jest.fn().mockReturnValue( '/' )
    } );

    const meta = renderer( ProjectsPage, {}, {
        DI,
        directives: {
            ...directives,
            hrefto
        }
    } );
    expect( storeMock.mock.calls ).toHaveLength( 1 );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'render errors', () => {
    const DI = createDI();
    const storeMock = jest.fn().mockReturnValue( Promise.reject() );
    DI.bind( 'store', {
        getProjects: storeMock
    } );
    DI.bind( 'router', {
        generate: jest.fn().mockReturnValue( '/' )
    } );

    const meta = renderer( ProjectsPage, {}, {
        DI,
        directives: {
            ...directives,
            hrefto
        }
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
} );
