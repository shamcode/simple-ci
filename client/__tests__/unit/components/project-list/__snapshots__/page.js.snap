// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render errors 1`] = `
Object {
  "Constructor": "ProjectsPage",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "dataLoaded": false,
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "hrefto": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
      "ref": [Function],
    },
    "errors": Array [],
    "projects": Array [],
    "types": Array [],
  },
  "html": "
<!--breadcrumbs-->
<!--if-->
<!--FlushPageDataLoading-->
<!--if-->
<!--PageContent-->
",
}
`;

exports[`renders correctly 1`] = `
Object {
  "Constructor": "ProjectsPage",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "dataLoaded": false,
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "hrefto": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
      "ref": [Function],
    },
    "errors": Array [],
    "projects": Array [],
    "types": Array [],
  },
  "html": "
<!--breadcrumbs-->
<!--if-->
<!--FlushPageDataLoading-->
<!--if-->
<!--PageContent-->
",
}
`;
